### PR TITLE
Feature: Added custom defender observation space.

### PR DESCRIPTION
**Pull request by Christian-Fisher**
Also added basic negative reward for defender.

Fixes:
is_defender_action_valid bug which caused outgoing rules to be ignored.

Signed-off-by: Christian-Fisher <christian.fisher@carleton.ca>

### CODE REVIEW

File: marlon/baseline_models/env_wrappers/defend_wrapper.py [hunk 1]
```diff
@@ -36,34 +42,21 @@ def __init__(self, cyber_env: CyberBattleEnv, max_timesteps=100, enable_action_p
...
+        """Creates an observation space for the defender. Consists of:
+        a list all node's infected status,
+        All nodes incoming firewall status,
+        all nodes outgoing firewall status,
+        all services statuses."""
+
+        # Calculate how many services there are, this is used to define the maximum number of services active at once.
+        for _, node in model.iterate_network_nodes(cyber_env.environment.network):
+            for _ in node.services:
+                self.num_services +=1
```
Comment (James-LG) [inline marlon/baseline_models/env_wrappers/defend_wrapper.py:54]:
```suggestion
        for _, node in model.iterate_network_nodes(cyber_env.environment.network):
            self.num_services += len(node.services)
```

File: marlon/baseline_models/env_wrappers/defend_wrapper.py [hunk 2]
```diff
@@ -36,34 +42,21 @@ def __init__(self, cyber_env: CyberBattleEnv, max_timesteps=100, enable_action_p
...
+        # All spaces are MultiBinary.
+        return spaces.Dict({'infected_nodes': spaces.MultiBinary(len(list(cyber_env.environment.network.nodes))),
```
Comment (James-LG) [inline marlon/baseline_models/env_wrappers/defend_wrapper.py:56]: Is the `len(list(cyber_env...))` required or can you just `len(cyber_env...)`?

File: marlon/baseline_models/env_wrappers/defend_wrapper.py [hunk 3]
```diff
@@ -108,13 +101,18 @@ def step(self, action) -> Tuple[Observation, float, bool, Dict[str, Any]]:
         attacker_action = self.cyber_env.sample_valid_action(kinds=[0, 1, 2])
         
         # Execute the step
-        observation, reward, done, info = self.cyber_env.step(attacker_action)
-        transformed_observation = self.transform_observation(observation)
+        attacker_observation, reward, done, info = self.cyber_env.step(attacker_action)
```
Comment (James-LG) [inline marlon/baseline_models/env_wrappers/defend_wrapper.py:104]: If a variable isn't used you can prepend an underscore so pylint doesn't complain about unused variables while maintaining code readability. E.g. `_attacker_observation`.

File: marlon/baseline_models/env_wrappers/defend_wrapper.py [hunk 4]
```diff
@@ -206,48 +204,62 @@ def service_exists(node_info: model.NodeInfo, service_from_action: int):
...
+    def get_blank_defender_observation(self):
+        """Creates a empty defender observation."""
+        obs = Defender_Observation(infected_nodes = [],
+                                    incoming_firewall_status=[],
+                                    outgoing_firewall_status=[],
+                                    services_status=[])
+        return obs
+
+    def observe(self) -> Defender_Observation:
+        """Gathers information directly from the environment to generate populate an observation for the defender agent to use."""
```
Comment (James-LG) [inline marlon/baseline_models/env_wrappers/defend_wrapper.py:221]: No dependency on the attacker's observation, nice! :+1:

**Review by James-LG — APPROVED** [review]

**Review by Chrisnguyen3 — APPROVED** [review]

**Pull request merged**